store/postgres/src/deployment_store.rs

```diff
@@ -1579,10 +1589,72 @@ impl DeploymentStore {
+
+        //check if all indexes are valid and recreate them
+        let namespace = site.namespace.as_str();
+        let mut conn = self.get_conn()?;
+        for table in dst.tables.values() {
+            for (index_name, create_query) in table.create_postponed_indexes() {
+                let table_name = table.name.clone();
+                // Version 1 - SQL generated with diesel
+                use pg_class as c;
+                use pg_index as x;
+                use pg_namespace as n;
+                let ns = diesel::alias!(pg_namespace as ns);
+                let (ca, ia) = diesel::alias!(pg_class as ca, pg_class as ia);
+                let q = x::table
+                    .inner_join(ca.on(ca.field(c::oid).eq(x::tabid)))
+                    .inner_join(ia.on(ia.field(c::oid).eq(x::indid)))
+                    .left_join(ns.on(ns.field(n::oid).eq(ca.field(c::namespace))))
+                    .filter(ns.field(n::name).eq(namespace))
+                    .filter(ca.field(c::name).eq(table_name.as_str()))
+                    .filter(ia.field(c::name).eq(index_name.as_str()))
+                    .filter(ca.field(c::kind).eq_any(vec!["r", "m", "p"]))
+                    .filter(ia.field(c::kind).eq_any(vec!["i", "I"]))
+                    .select((x::isvalid, x::isready))
+                    .get_results::<(bool, bool)>(&mut conn)?
+                    .into_iter()
+                    .collect::<Vec<(bool, bool)>>();
+                println!("SQL: {:?}", q);
+                // Version 2 - A string SQL:
+                let query = r#"
+                    SELECT  x.indisvalid           AS isvalid,
+                            x.indisready           AS isready
+                    FROM pg_index x
+                            JOIN pg_class c ON c.oid = x.indrelid
+                            JOIN pg_class i ON i.oid = x.indexrelid
+                            LEFT JOIN pg_namespace n ON n.oid = c.relnamespace
+                    WHERE (c.relkind = ANY (ARRAY ['r'::"char", 'm'::"char", 'p'::"char"]))
+                    AND (i.relkind = ANY (ARRAY ['i'::"char", 'I'::"char"]))
+                    AND (n.nspname = $1)
+                    AND (c.relname = $2)
+                    AND (i.relname = $3);"#;
+                let ii_vec = sql_query(query)
+                    .bind::<Text, _>(namespace)
+                    .bind::<Text, _>(table_name)
+                    .bind::<Text, _>(index_name.clone())
+                    .get_results::<IndexInfo>(&mut conn)?
+                    .into_iter()
+                    .map(|ii| ii.into())
+                    .collect::<Vec<IndexInfo>>();
+                // TODO: pick one of the above
+                assert!(ii_vec.len() <= 1);
+                // Check if the index is valid. If either isvalid or isready flag of pg_index table
+                // isn't true, drop it and recreate it.
+                if ii_vec.len() == 0 || !ii_vec[0].isvalid || !ii_vec[0].isready {
```

Comment:
Criteria for recreating the index is that either `inisvalid` or `indisready`  is false from `pg_index` table. If the respective row is completely missing I'd skip dropping the old index and proceed with creating the new one directly. WDYT @lutter

Reply:
I guess we decided to pick the raw SQL string.